### ISSUE TITLE
Get rid of sys.setdefaultencoding and move to Unicode by default

### ISSUE THREAD

**Issue by benjaoming** [learningequality/ka-lite#3434]
In `kalite.settings` there's the following block of code (some modifications because of the setuptools installer)

``` python
# set the default encoding
# OK, so why do we reload sys? Because apparently sys.setdefaultencoding
# is deleted somewhere at startup. Reloading brings it back.
# see: http://blog.ianbicking.org/illusive-setdefaultencoding.html

# benjaoming: We need to somehow get rid of this as it's discouraged
# http://stackoverflow.com/questions/3828723/why-we-need-sys-setdefaultencodingutf-8-in-a-py-script
# http://ziade.org/2008/01/08/syssetdefaultencoding-is-evil/
try:
    DEFAULT_ENCODING = DEFAULT_ENCODING
except NameError:
    from django.conf.settings import DEFAULT_ENCODING

if sys.getdefaultencoding() != DEFAULT_ENCODING:
    reload(sys)
    sys.setdefaultencoding(DEFAULT_ENCODING)
```

The motivation behind this is memory optimization. Experience reported by @aronasorman and @66eli77 says that memory efficiency is ~1.5 and this is much crucial for Raspberry Pi execution.

Indeed, this is also true... but it's very sensitive to the string size. This is pretty crazy, because the [Python 2 docs](https://docs.python.org/2/howto/unicode.html#encodings) say the following...

> UTF-8 is fairly compact; the majority of code points are turned into two bytes, and values less than 128 occupy only a single byte.

Let's do 3 characters of ascii....

```
>>> from sys import getsizeof
>>> getsizeof("123")
40
>>> getsizeof(u"123")
64
```

But then, lets do a lot more...

```
>>> getsizeof("123" * 1000)
3037
>>> getsizeof(u"123" * 1000)
12052
```

Now the unicode type is consuming 4x as much space!? Let's do something with unicode characters...

```
>>> getsizeof("®€łœßĸ“đłłđ®" * 1000)
26037
>>> getsizeof(u"®€łœßĸ“đłłđ®" * 1000)
48052
```

Now it's less than 2x as much...

The problem will go away in Python 3.. as I understand it, the unicode (which is the way that strings will be treated by default), is also optimized for choosing a better codepoint something something...

```
Python 3.4.0 (default, Apr 11 2014, 13:05:11) 
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sys import getsizeof
>>> getsizeof("123")
52
>>> getsizeof("123" * 1000)
3049
```

As you see, Python 3, perfectly handles UTF-8 encoding.

Now, here's the bible of all this unicode stuff:

https://docs.python.org/2/howto/unicode.html
# What to do!!?

I suggest moving everything to Py2+3 compatibility, which will mean using `from __future__ import unicode_literals`. From my testing, this does **not** solve the memory issues, but it's a necessary step to start including Python 3 compatibility.

This would leave us with a memory problem, but we have to solve it by reducing our dependency on loading big JSON files in memory. This is what databases are used for :D

Another issue solved by forcing the default encoding to `UTF-8` has been to avoid explicitly encoding/decoding `str` objects. This should also be solved by moving to `unicode_literals` on Python 2.

**Comment by jamalex** [comment]
Definitely want to just be using Unicode eventually. But we can't wait for Python 3 for that, since we aim to be backwards compatible, and that's *especially* important on the same devices that are low memory.

Moving away from having all that JSON loaded in memory is critical. Would be good to get some benchmarking started on DB-based approaches. Would probably be good to have this stuff in a different DB backend than the user data, for several reasons (e.g. sqlite's low concurrency, and the use of copying a DB file as a method of backup).

**Comment by benjaoming** [comment]
>  But we can't wait for Python 3 for that

It's been there for quite a while :D But I wouldn't be scared of Py2+3 compatibility, it can be done, although we'll see a couple of issues maintaining 2.6.5, too... but this problem goes for Django as well as any other Django 2+3 apps out there (and there are loads).

Removing the unicode hacks would be one of these first steps, dodging whatever performance hits we've taken after that, would be the second step... and maybe we could start by just using `unicode_literals` without actually being fully python3 compatible. But at the end of it, we'd most likely see a performance benefit... both in Python 2 and 3.

It should be easy to see performance improve, because there's only *so much* data being loaded directly into memory at the moment. Depending on the types of lookups and data structures, we may even see the database (SQLite) reducing computational complexities.

For filtering, sorting, and joining related data, a database should be quite superior :)

> Would probably be good to have this stuff in a different DB backend than the user data

Yes, database routing can do this. Using SQLite is just an option, but at the end of the day, this is a project setting that can be changed according to needs.. so prime goal would be fixing the unicode hacks and the ways we store data... and then as a result of that we can look at project settings in relation to benchmarks.. I mean, would we need benchmarking to decide that data has to go in its own database according to what the project settings define?

We probably won't get around data (topic data, assessment items etc) being distributed by JSON.

Road map could be....

1. Fix unicode hacks (introduce unicode_literals) => Performance decrease (memory x 1.5 for the string handling parts)
1. Store in a separate db instead of JSON and load-on-demand => Performance increase (memory x 0.0 for the string handling parts!!!)
1. Introduce Py2+3 fully and start supporting Python 3. At this point, we don't have to do the whole unicode thing as it was already done in the first step. => Performance benefit for all Python 3, as we know that it has a much more memory efficient way of using strings, and it can run on Raspberry Pi.

Also as a little secondary note, removing the encoding hacks and being 100% unicode is pretty much a prerequisite for moving data to the database. Imagine people doing upgrades and having encoding issues in their DB because we tried to move from one encoding to another.. ouch.

**Comment by MCGallaspy** [comment]
We're starting to address this. See, for example #3639.

**Comment by benjaoming** [comment]
@MCGallaspy Can you give an update as to how many in-memory strings we have to deal with by now? Is it safe to move to UTF-8 in 0.15 ?

**Comment by aronasorman** [comment]
@benjaoming you can try switching to that now, and see what the difference in memory usage is. FYI mine stabilizes at about ~300MB in the current develop branch.

**Comment by MCGallaspy** [comment]
No, I don't think this will be addressed in 0.15. Aside from the assessment items, the other big use of in-memory strings is the so-called content data (the tree-structure of topics, videos and exercises).

**Comment by rtibbles** [comment]
Given that we are moving these strings out of memory now, hopefully this can be addressed in 0.16 still.

**Comment by benjaoming** [comment]
KA Lite will not support Python 3, and I've added a comment in the code regarding this issue. Since we don't intend to fix it, keeping the issue open won't help.

Furthermore, the uncertainty about string handling which would be introduced by changing `setdefaultencoding` is substantial, so doesn't seem terribly appealing given the modest testing resources we would have for a new release.
